src/HTTP/Duration/CallbackDuration.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+/**
+ * ******************************************************************
+ *
+ * This file is part of ILIAS, a powerful learning management system
+ * published by ILIAS open source e-Learning e.V.
+ *
+ * ILIAS is licensed with the GPL-3.0, see
+ * https://www.gnu.org/licenses/gpl-3.0.en.html
+ * You should have received a copy of said license along with the
+ * source code, too.
+ *
+ * If this is not the case or you just want to try ILIAS, you'll find
+ * us at:
+ *      https://www.ilias.de
+ *      https://github.com/ILIAS-eLearning
+ *
+ * ******************************************************************
+ */
+
+namespace ILIAS\HTTP\Duration;
+
+use Closure;
+
+/**
+ * @author Thibeau Fuhrer <thibeau@sr.solutions>
+ */
+class CallbackDuration extends Duration
+{
+    /**
+     * @return mixed|null
+     */
+    public function stretch(callable $callback)
+    {
+        $start_time = microtime(true);
+        $halted = true;
+
+        register_shutdown_function(static function () use (&$halted) {
+            if ($halted && (DEVMODE || DEBUG)) {
```

Comment:
I would suggest to remove the condition on `DEVMODE` or `DEBUG`. If this `not $halted`, there is a serious problem since something prevents this security measure from working. This should not happen in any sane production installation, hence `\LogicException`. IF this happens, this should me communicated loud and clear.